components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/RequestObjectValidatorImpl.java

```diff
@@ -74,14 +76,101 @@ public boolean isSigned(RequestObject requestObject) {
-        boolean isVerified = isSignatureVerified(jwt, certificate);
+        try {
+             certificate =
+                    getCertificateForAlias(oAuth2Parameters.getTenantDomain(), oAuth2Parameters.getClientId());
+        } catch (RequestObjectException e) {
+            log.warn("Error retrieving public certificate for service provider checking whether a jwks endpoint " +
+                    "is configured for the service provider ");
+            if (log.isDebugEnabled()) {
+
+                log.debug("Error retrieving public certificate for service provider checking whether a jwks endpoint " +
+                        "is configured for the service provider: " + e);
+            }
+        }
+        if (certificate == null) {
+            if (log.isDebugEnabled()) {
+
+                log.debug("Public certificate not configured for Service Provider with " +
+                        "client_id: " + oAuth2Parameters.getClientId() + " of tenantDomain: " + oAuth2Parameters
+                        .getTenantDomain() + ". Fetching the jwks endpoint for validating request object");
+            }
+            String jwksUri = getJWKSEndpoint(oAuth2Parameters);
+            isVerified = isSignatureVerified(jwt, jwksUri);
+        } else {
+            if (log.isDebugEnabled()) {
+
+                log.debug("Public certificate configured for Service Provider with " +
+                        "client_id: " + oAuth2Parameters.getClientId() + " of tenantDomain: " + oAuth2Parameters
+                        .getTenantDomain() + ". Using public certificate  for validating request object");
+            }
+            isVerified = isSignatureVerified(jwt, certificate);
+        }
         requestObject.setIsSignatureValid(isVerified);
         return isVerified;
     }
 
+    /**
+     * Fetch JWKS endpoint using OAuth2 Parameters.
+     *
+     * @param oAuth2Parameters  oAuth2Parameters
+     */
+    private String getJWKSEndpoint(OAuth2Parameters oAuth2Parameters) throws RequestObjectException {
+
+        String jwksUri = StringUtils.EMPTY;
+        ServiceProviderProperty[] spProperties;
+        try {
+            spProperties = OAuth2Util.getServiceProvider(oAuth2Parameters.getClientId())
+                    .getSpProperties();
+        } catch (IdentityOAuth2Exception e) {
+            throw new RequestObjectException("Error while getting the service provider for client ID " +
+                    oAuth2Parameters.getClientId(), OAuth2ErrorCodes.SERVER_ERROR, e);
+        }
+
+        if (spProperties != null) {
+            for (ServiceProviderProperty spProperty : spProperties) {
+                if (Constants.JWKS_URI.equals(spProperty.getName())) {
+                    jwksUri = spProperty.getValue();
+                    if (log.isDebugEnabled()) {
+                        log.debug("Found jwks endpoint " + jwksUri + "for service provider with client id " +
```

Comment:
Add a space between the logged jwks uri and the start of the next string message

Reply:
done